File: 0031-next-permutation/0031-next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        for ind in range(len(nums)-1, -1, -1):
            if ind == 0:
                nums.sort()
                return
            elif nums[ind-1] < nums[ind]:
                pivot_index = ind - 1
                pivot_element = nums[ind - 1]
                break
        
        for index in range(len(nums)-1, -1, -1):
            if nums[index] > pivot_element:
                nums[index], nums[pivot_index] = nums[pivot_index], nums[index]
                break
        
        left_pointer = pivot_index + 1
        right_pointer = len(nums) - 1

        while left_pointer < right_pointer:
            nums[left_pointer], nums[right_pointer] = nums[right_pointer], nums[left_pointer]
            left_pointer += 1
            right_pointer -= 1

File: 0031-next-permutation/0031-next-permutation.java
class Solution {
        public void nextPermutation(int[] nums){
        int leftPointer = 0;
        int rightPointer = nums.length -1;
        int pivotElement = 0;
        int pivotIndex = 0;

        for (int index = nums.length-1; index > -1; index--) {
            if (index == 0){
                Arrays.sort(nums);
                return;
            } else if (nums[index-1] < nums[index]) {
                pivotIndex = index - 1;
                leftPointer = index;
                pivotElement = nums[index-1];
                break;
            }
        }

        for (int ind = nums.length-1; ind > -1; ind--) {
            if (nums[ind] > pivotElement){
                swap(nums, ind, pivotIndex);
                break;
            }
        }

        while(leftPointer < rightPointer){
            swap(nums, leftPointer, rightPointer);
            leftPointer ++;
            rightPointer --;
        }

    }

    public void swap(int[] array, int index1, int index2){
        int tempInt = array[index1];
        array[index1] = array[index2];
        array[index2] = tempInt;
    }
}